common/flatpak-dir.c

```diff
@@ -14628,3 +14628,345 @@ flatpak_dir_delete_mirror_refs (FlatpakDir    *self,
+  for (guint i = 0; i < len; i++)
+    {
+      if (g_strv_contains (refs_to_exclude, full_refs[i]))
+        {
+          g_free (full_refs[i]);
+          full_refs[i] = full_refs[len - 1];
+          full_refs[len - 1] = NULL;
+          len--;
+          i--;
+        }
+    }
+}
+
+/* See the documentation for
+ * flatpak_installation_list_unused_refs_with_options().
+ * The returned pointer array is transfer full. */
+char **
+flatpak_dir_list_unused_refs_with_options (FlatpakDir         *self,
+                                           const char         *arch,
+                                           GHashTable         *metadata_injection,
+                                           const char * const *refs_to_exclude,
+                                           gboolean            filter_by_eol,
+                                           GCancellable       *cancellable,
+                                           GError            **error)
+{
+  g_autoptr(GHashTable) refs_hash = NULL;
+  g_autoptr(GPtrArray) refs =  NULL;
+  g_auto(GStrv) app_refs = NULL;
+  g_auto(GStrv) runtime_refs = NULL;
+  g_autoptr(GHashTable) used_refs = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+  g_autoptr(GHashTable) used_runtimes = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+  int i;
+
+  if (!flatpak_dir_list_refs (self, "app", &app_refs, cancellable, error))
+    return NULL;
+
+  if (!flatpak_dir_list_refs (self, "runtime", &runtime_refs, cancellable, error))
+    return NULL;
+
+  if (refs_to_exclude != NULL)
+    {
+      prune_excluded_refs (app_refs, refs_to_exclude);
+      prune_excluded_refs (runtime_refs, refs_to_exclude);
+    }
+
+  refs_hash = g_hash_table_new (g_str_hash, g_str_equal);
+  refs = g_ptr_array_new_with_free_func (g_free);
+
+  /* For each app, note the runtime, sdk, and related refs */
+  find_used_refs_for_apps (self, NULL, app_refs, arch, metadata_injection, used_runtimes, used_refs);
+
+  /* If @self is a system installation, also check the per-user installation
+   * for any apps there using runtimes in the system installation or runtimes
+   * there with sdks or extensions in the system installation. Only do so if
+   * the per-user installation exists; it wouldn't make sense to create it here
+   * if not.
+   */
+  if (!flatpak_dir_is_user (self))
+    {
+      g_autoptr(GFile) user_base_dir = flatpak_get_user_base_dir_location ();
+
+      if (g_file_query_exists (user_base_dir, cancellable))
+        {
+          g_autoptr(FlatpakDir) user_dir = flatpak_dir_get_user ();
+          g_auto(GStrv) user_app_refs = NULL;
+          g_auto(GStrv) user_runtime_refs = NULL;
+          g_autoptr(GHashTable) user_runtimes = g_hash_table_new_full (g_str_hash, g_str_equal, NULL, NULL);
+
+          if (!flatpak_dir_list_refs (user_dir, "app", &user_app_refs, cancellable, error))
+            return NULL;
+
+          /* metadata_injection is NULL because it's not for this installation */
```

Comment:
Is this right? Metadata_injection applies to the installation directory, which in this case is "self" (and later "system_dir"). Later find_used_refs_for_apps() uses the metadata injection in two places, to load the app itselfs metadata, and to load the metadata for the app dependencies. We should *not* use the injection for the app metadata in this case, but we *should* use it for the dependencies.